[Fiber] Support SVG by adding a container stack

gaearon opened the pull request:
I added a test for SVG. For it to pass, we need to keep track of the closest host container (e.g. `svg` or `foreignobject`) on the stack. <s>Since we need the stack anyway, I also used it to append initial children in the begin phase instead of the complete phase. I am using two separate stacks: one for "container" nodes (such as `svg` and `foreignobject`) and another one for any "new" nodes so that we can add children to them immediately.</s> Changed to use a single container stack per review comments. I can do the parent stack in a follow up.

gaearon commented on scripts/fiber/tests-passing.txt:
```diff
@@ -633,11 +586,7 @@ src/renderers/dom/shared/__tests__/ReactDOMInvalidARIAHook-test.js
-* updates a mounted text component in place
-* can be toggled in and out of the markup
-* can reconcile text from pre-rendered markup
+* creates elements with svg namespace inside svg tag
```
The only passing test is here.

gaearon commented:
@sebmarbage This is sort of ready for review. Portals might be broken but I have a todo for that.

gaearon commented:
Should work with portals too now.

gaearon commented:
![](https://d17oy1vhnax1f7.cloudfront.net/items/2D3O1a0k1q0Y2Y1b3o3x/Screen%20Recording%202016-11-25%20at%2022.01.gif?v=5d3ffc87)

gaearon commented:
Note that I haven't tried unifying this with portals yet. I'm not sure if it's helpful or not.

sebmarkbage commented:
Any way we can split the container piece from the parent piece (that creates in beginWork)? The second one is a hot path and needs further evaluation.

gaearon commented:
Removed the parent stack and moved initial insertions back into complete phase.

We still need to create the instances in the begin phase so that we have access to root container instance in `ReactDOMFiberComponent.createElement()` and can determine the namespace and owner document.

gaearon commented on src/renderers/shared/fiber/ReactFiberBeginWork.js:
```diff
@@ -453,8 +497,15 @@ module.exports = function<T, P, I, TI, C>(
       case HostText:
-        // Nothing to do here. This is terminal. We'll do the completion step
-        // immediately after.
+        const newText = workInProgress.pendingProps;
```
Technically didn't have to move this one but did this for consistency.
Can move it back too.

sebmarkbage commented on src/renderers/dom/fiber/ReactDOMFiber.js:
```diff
@@ -107,14 +119,19 @@ var DOMRenderer = ReactFiberReconciler({
+    // and use tagName (which is upper case for HTML DOM elements). Or we could
+    // let the renderer "normalize" the fiber type so we don't have to read
+    // the type from DOM. However we need to remember SVG is case-sensitive.
+    var tag = domElement.tagName.toLowerCase();
```
This is likely going to be an issue to do for every update. We might want to consider just banning non-lowercase uses instead.

gaearon replied:
I can omit this change if it's better to discuss separately.

gaearon replied:
Also, we already doing this in `commitUpdate()` before my PR. So I don't see why doing this during mounting is worse. (We should fix both?)

sebmarkbage commented:
The creation order is only needed if we want to read the namespace from the DOM. We don't have to design it that way. We can pass a flag or something instead. It might be more efficient than touching DOM accessor methods anyway and doesn't lock us in to this creation order. (Even though we might want that anyway.)

Will review further today. The reason it is taking time is that I want to a deep review and consideration because this change is a compromise only for the purpose of letting users avoid switching renderers in the rare cases you switch modes. It is also very specific only to the DOM renderer. Anything that adds perf overhead or locks us into a particular creation phase seems like a big tradeoff.

gaearon commented on src/renderers/shared/fiber/__tests__/ReactIncrementalSideEffects-test.js:
```diff
@@ -517,22 +517,22 @@ describe('ReactIncrementalSideEffects', () => {
     }
     ReactNoop.render(<Foo tick={0} idx={0} />);
-    ReactNoop.flushDeferredPri(40 + 25);
+    ReactNoop.flushDeferredPri(40 + 20);
```
Less cycles, isn't it? I'm not sure. Should I look into this?

gaearon commented:
>letting users avoid switching renderers in the rare cases you switch modes

I don’t understand this part. What case are you talking about?

gaearon commented:
I’m looking into tracking namespaces separately so that we don’t need to change when we create.

gaearon commented:
@sebmarkbage Should be ready for another review. Now only tracking the namespace, instances are created like before.

sebmarkbage commented:
> > letting users avoid switching renderers in the rare cases you switch modes

> I don’t understand this part. What case are you talking about?

I mean that we could also just make people write:

```js
return <div>{ReactDOM.createSVGPortal(
  <svg>
     <path />
  </svg>
)}</div>;
```

I'm merely making the point that this whole thing is mostly just a convenience. Any additional cost in file size and performance needs to take that tradeoff into account.

But it seems like we can get it pretty small.

gaearon commented on src/renderers/dom/fiber/ReactDOMFiberComponent.js:
```diff
@@ -452,44 +457,39 @@ function updateDOMProperties(
+      }
+    }
+    if (parentNamespace === SVG_NAMESPACE && type === 'foreignObject') {
+      return null;
```
Fixed.

sebmarkbage replied:
Can you add an assertion that the `<svg />` tag itself has the SVG namespace?

sebmarkbage commented:
This gave me an idea. Feel free to ignore if you're tired of exploring my whims :) and I can take a look myself.

1) What if we tagged a new type of work HostContextComponent or something. If the type is a context provider, it gets tagged as such. That way we don't need to keep the stack of fibers around to know if we need to pop. HostContextComponent always pushes and pops. It also eliminates any extra work from the normal host component.

2) What if just assumed that when you pop from MathML or SVG, you pop into HTML and when you pop from SVG you pop to HTML? That way we don't need a stack at all.

gaearon commented:
I’ll try that. 😄

gaearon commented:
Closing as we decided to go with a cheaper approach.
